mongodb_exporter.go

```diff
@@ -42,47 +41,42 @@ func defaultMongoDBURI() string {
-	tlsDisableHostnameValidationF = flag.Bool("mongodb.tls-disable-hostname-validation", false, "Disable hostname validation for server connection.")
-	maxConnectionsF               = flag.Int("mongodb.max-connections", 1, "Max number of pooled connections to the database.")
-	testF                         = flag.Bool("test", false, "Check MongoDB connection, print buildInfo() information and exit.")
-
-	socketTimeoutF = flag.Duration("mongodb.socket-timeout", 3*time.Second, "Amount of time to wait for a non-responding socket to the database before it is forcefully closed.\n"+
-		"    \tValid time units are 'ns', 'us' (or 'µs'), 'ms', 's', 'm', 'h'.")
-	syncTimeoutF = flag.Duration("mongodb.sync-timeout", time.Minute, "Amount of time an operation with this session will wait before returning an error in case\n"+
+		"    \tIf not provided: System default CAs are used.").Default("").String()
+	tlsDisableHostnameValidationF = kingpin.Flag("mongodb.tls-disable-hostname-validation", "Disable hostname validation for server connection.").Bool()
+	maxConnectionsF               = kingpin.Flag("mongodb.max-connections", "Max number of pooled connections to the database.").Default("1").Int()
+	testF                         = kingpin.Flag("test", "Check MongoDB connection, print buildInfo() information and exit.").Bool()
+
+	socketTimeoutF = kingpin.Flag("mongodb.socket-timeout", "Amount of time to wait for a non-responding socket to the database before it is forcefully closed.\n"+
+		"    \tValid time units are 'ns', 'us' (or 'µs'), 'ms', 's', 'm', 'h'.").Default("3s").Duration()
+	syncTimeoutF = kingpin.Flag("mongodb.sync-timeout", "Amount of time an operation with this session will wait before returning an error in case\n"+
 		"    \ta connection to a usable server can't be established.\n"+
-		"    \tValid time units are 'ns', 'us' (or 'µs'), 'ms', 's', 'm', 'h'.")
+		"    \tValid time units are 'ns', 'us' (or 'µs'), 'ms', 's', 'm', 'h'.").Default("1m").Duration()
 
 	// FIXME currently ignored
 	// enabledGroupsFlag = flag.String("groups.enabled", "asserts,durability,background_flushing,connections,extra_info,global_lock,index_counters,network,op_counters,op_counters_repl,memory,locks,metrics", "Comma-separated list of groups to use, for more info see: docs.mongodb.org/manual/reference/command/serverStatus/")
-	enabledGroupsFlag = flag.String("groups.enabled", "", "Currently ignored")
+	enabledGroupsFlag = kingpin.Flag("groups.enabled", "Currently ignored").Default("").String()
 )
 
 func main() {
-	flag.Usage = func() {
-		fmt.Fprintf(os.Stderr, "%s %s exports various MongoDB metrics in Prometheus format.\n", os.Args[0], version.Version)
-		fmt.Fprintf(os.Stderr, "Usage: %s [flags]\n\n", os.Args[0])
-		fmt.Fprintf(os.Stderr, "Flags:\n")
-		flag.PrintDefaults()
-	}
-	flag.Parse()
+	kingpin.CommandLine.Help = fmt.Sprintf("%s %s exports various MongoDB metrics in Prometheus format.\n", os.Args[0], version.Version)
+	kingpin.Parse()
```

Comment:
```go
kingpin.HelpFlag.Short('h')
```

Plus add `kingpin.Version(version.Version)` or something like that.

Reply:
I'll add short help flag, but there is no reason to add `kingpin.Version(version.Version)` cause now we have extended version info right here https://github.com/percona/mongodb_exporter/blob/5fa2ea81292eb750161ca79eae66fce5a168e668/mongodb_exporter.go#L110
If I use it, we will see only `X.Y.Z` in output and there is no way to add custom version output.

Reply:
That `if` should be removed – kingpin will handle that flag itself.
In the future, we will switch to https://github.com/percona/pmm/blob/PMM-2.0/version/release.go like that: https://github.com/percona/pmm-agent/pull/12/commits/9e5b9a1ead40d76afa7651e65ce641d71bdf04e8

Reply:
See https://jira.percona.com/browse/PMM-2567, https://jira.percona.com/browse/PMM-1168, and https://jira.percona.com/browse/PMM-3513.

Reply:
I thought `version` needs to bee, like `0.6.3` in context of `kingpin.Application`. It was small misunderstanding, so I'll remove it.